Add user-level email address visibility setting.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
This PR adds user-level email address visibility setting (and also the realm-level default for new users) and removes the existing realm setting.

This PR replaces #20102 since it had a lot of clutter and was difficult to review.
Commits and changes are same.
<!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

sahil839 commented on zerver/lib/events.py:
```diff
@@ -844,9 +847,17 @@ def _draft_update_action(i: int) -> None:
+                if (
+                    person["email_address_visibility"]
+                    != UserProfile.EMAIL_ADDRESS_VISIBILITY_EVERYONE
+                ):
```
This change still needs to be discussed. We were thinking of not including the `email_address_visibility` value in the person objects of this event and for `realm_user/add` event above. One solution is to just check the `delivery_email` field, but in that case we would set `client_gravatar` as `True` for a user who has set `email_address_visibility` setting to `Admins only` and the acting user is admin and we would not sent avatar_url field in such case and expect clients to compute it from email.

This was not the case previously, as we did not send the`avatar_url` field only when email address visibility was set to `everyone`, i.e. the `avatar_url` field was sent to admin even if the visibility was set to `Admins only`. I am not sure whether to do this change or not as [this comment](https://github.com/zulip/zulip/blame/aa5536fe3046d76813015dd7db83297a213ca03a/zerver/tests/test_users.py#L361) mentions that we would need some changes for mobile app clients.

timabbott replied:
Has this been discussing in an "api design" thread? If not, please start one and let's get the mobile team's opinion on it.

timabbott replied:
Posted about it here: https://chat.zulip.org/#narrow/stream/378-api-design/topic/email.20address.20visibility/near/1497722.

sahil839 commented:
Left one comment above. A couple of more points -
- We still show fake emails in search input box. There was some discussion to instead show name if the real email is not visible and I tried doing it but it resulted in some complex changes (#20282 ). There is no problem in typeaheads as we have changed them to show user pills now. Also, note that we show fake emails in search input box even if the real email is available since the visibility setting is not set to `everyone` and we consider the `email` field always and not `delivery_email` field.
- We discussed [here](https://github.com/zulip/zulip/pull/20102#issuecomment-990239715) that we want to add user ID to user deactivation modal. We need to discuss the design but I am mentioning it here so that it does not go unnoticed.

sahil839 commented:
Adding a GIF above for the first bullet in above comment. Here the user actually can access Cordelia's real email but the search box still has fake email.
![search-email](https://user-images.githubusercontent.com/35494118/191014956-0d6a5e04-f791-49bd-8149-64dedc00836e.gif)

sahil839 commented:
@timabbott opened this PR as replacement of #20102.

sahil839 commented:
We can still do some work on live-updating the email at various places in UI like admin users list, stream subscribers list, etc. We also do not live-update email at these places when user changes their email. I am not sure how quick it will be to do it, so probably better as a separate project.

alya commented:
Thanks! The only thing I noticed in manual testing is that we should probably display the user's real email when they search for themselves in the top search bar (e.g., `from:me`).

sahil839 commented:
That's the same issue as it is there with other users. We show the `email` field of the user object (which is not the real email in case user has visibility other than everyone) and that is used to execute the search. I think it would better to instead work on changing the search to pills which I have started working on.

alya commented:
Ah, OK, that's all right, then. @timabbott ready for your review.

timabbott commented on zerver/lib/users.py:
```diff
@@ -402,10 +402,10 @@ def can_access_delivery_email(
 
-    if email_address_visibility == Realm.EMAIL_ADDRESS_VISIBILITY_MODERATORS:
+    if email_address_visibility == UserProfile.EMAIL_ADDRESS_VISIBILITY_MODERATORS:
         return user_profile.is_realm_admin or user_profile.is_moderator
```
This seems to not implement `EMAIL_ADDRESS_VISIBILITY_MEMBERS`?

sahil839 replied:
I planned to add it separately as it was not in the realm-wide setting. But added it now, since it was not much work to add it.

timabbott commented on zerver/lib/test_helpers.py:
```diff
@@ -199,12 +200,21 @@ def stdout_suppressed() -> Iterator[IO[str]]:
+            "email_address_visibility",
+            UserProfile.EMAIL_ADDRESS_VISIBILITY_EVERYONE,
+            acting_user=None,
+        )
```
This function seems totally crazy -- it looks like it is primarily called as the first step in a test ... which should be a situation where it's not going to do anything, because we're starting fresh after rolling back the transaction for the previous test. So I think with likely minimal adjustment, we can delete it and just remove it from all its callers.

That can be a separate PR or a prep commit in this one depending what feels convenient.

sahil839 replied:
By default the `email_address_visibility` for all the users is set to `ADMINS_ONLY` for the development and test database.

timabbott replied:
I see. I think `ADMINS_ONLY` being the default makes sense - we want to test general code paths with the possibility that users don't have access to email addresses, since that's the more complex case.

The solution is probably to just rename this function and/or give it a docstring explaining what we're doing and why.

timabbott commented on zerver/lib/message.py:
```diff
@@ -374,6 +374,13 @@ def finalize_payload(
+            # clients cannot compute gravatars, so we force-set it to false.
+            # If we plumbed the current user's role, we could allow client_gravatar=True
+            # here if the current user's role has access to the target user's email address.
+            client_gravatar = False
```
This is messy, but I think probably the best solution available. In the future, I think we're going to want to just move off Gravatar to our own procedural generator that is a function of immutable details like the user ID and realm ID, or perhaps a random seed, so I'm not sure it's worth cleaning this up really.

timabbott commented on zerver/lib/events.py:
```diff
@@ -772,8 +772,13 @@ def _draft_update_action(i: int) -> None:
-                person["avatar_url"] = None
+            if person["email_address_visibility"] != UserProfile.EMAIL_ADDRESS_VISIBILITY_EVERYONE:
+                client_gravatar = False
+            del person["email_address_visibility"]
```
I suppose we have tests that verify `email_address_visibility` won't appear in the output event.

I'll also note that we could potentially have this code call a variant of `can_access_email_address`, since we do have the calling `user_profile` and thus their role.

sahil839 replied:
I think by "output event" you meant the user object in the state that is added/update using this event. Yes, the tests verify this. To be clear, the `email_address_visibility` is still present in the event sent to client as of now.

timabbott replied:
Yeah, makes sense. We can resolve this item since whether `email_address_visibilty` should be in here is being discussed in the "api design" thread.

timabbott commented on zerver/lib/create_user.py:
```diff
@@ -166,6 +169,7 @@ def create_user(
         tos_version,
         timezone,
         default_language,
+        email_address_visibility=realm_user_default.email_address_visibility,
```
This is OK for integrating this PR. But probably after integrating this, we'll want to consider a feature to ask the email address visibility question to users during account creation (with the default pulled from `RealmUserDefault`). @alya FYI -- it'd be reasonable to do the #design discussion and/or issue write-up for that soon.

sahil839 replied:
Yeah, I think we discussed about this before.

timabbott commented on zerver/lib/create_user.py:
```diff
@@ -89,6 +89,7 @@ def create_user_profile(
     timezone: str,
     default_language: str = "en",
     tutorial_status: str = UserProfile.TUTORIAL_WAITING,
+    email_address_visibility: Optional[int] = UserProfile.EMAIL_ADDRESS_VISIBILITY_EVERYONE,
```
This is an `int`, not `Optional`, right?

sahil839 replied:
Yes, updated the type to be `int` and also made it mandatory kwarg.

timabbott commented on zerver/actions/create_realm.py:
```diff
@@ -211,7 +210,19 @@ def do_create_realm(
+            # Email address of users should be initially visible to admins only.
+            realm_default_email_address_visibility = (
+                RealmUserDefault.EMAIL_ADDRESS_VISIBILITY_ADMINS
+            )
```
This can be a follow-up, but @alya we should probably decide which organization types we'd like to have what email address visibility in newly create organizations.

(I don't think it's particularly important for merging this PR, since it's just an initial setting)

timabbott commented on zerver/actions/create_user.py:
```diff
@@ -317,7 +317,7 @@ def notify_created_user(user_profile: UserProfile) -> None:
 
     if user_ids_without_real_email_access:
-        del person["delivery_email"]
+        person["delivery_email"] = None
```
It probably would be a nice future refactor to just call `format_user_row` twice with different parameters, rather than doing this surgery. I'm pretty sure that function doesn't have any real cost, and one wouldn't have to think about this surgery too closely. But I don't think that's a blocker for merging this.

sahil839 replied:
Will look at this.

timabbott commented on zerver/actions/user_settings.py:
```diff
@@ -52,6 +52,42 @@ def send_user_email_update_event(user_profile: UserProfile) -> None:
+        if can_access_delivery_email(
+            user, user_profile.id, old_visibility_setting
+        ) != can_access_delivery_email(user, user_profile.id, visibility_setting):
+            if can_access_delivery_email(user, user_profile.id, visibility_setting):
```
This would be a lot more readable with more variables -- just `could_access_previously = ...`; `can_access_now = ...`, and then using those for the if/else statements here. 

Also avoids a second call to `can_access_delivery_email`; that's not a big deal as the function is cheap.

sahil839 replied:
Updated the PR with these new variables.

timabbott commented on zerver/actions/user_settings.py:
```diff
@@ -52,6 +52,42 @@ def send_user_email_update_event(user_profile: UserProfile) -> None:
+) -> None:
+    active_users = user_profile.realm.get_active_users()
+    delivery_email_visible_user_ids = []
+    delivery_email_not_visible_user_ids = []
```
`delivery_email_now_visible_user_ids` and `delivery_email_now_invisible_user_ids` might be better variable names, given that we're not doing anything special for ones that haven't changed?

sahil839 replied:
Updated.

timabbott commented on zerver/actions/user_settings.py:
```diff
@@ -63,12 +99,19 @@ def do_change_user_delivery_email(user_profile: UserProfile, new_email: str) ->
+    active_users = user_profile.realm.get_active_users()
+    delivery_email_visible_user_ids = [
+        user.id
+        for user in active_users
```
Something like `for active_user in active_users` might be a nicer variable name -- I don't like ever having two variables `user_profile` and `user` in the same function. Eventually we're probably going to rename `UserProfile` to `User` and try to squash all of those.

sahil839 replied:
Updated this.

timabbott commented on zerver/lib/users.py:
```diff
@@ -483,6 +486,7 @@ def format_user_row(
             client_gravatar=client_gravatar,
         )
 
+    result["delivery_email"] = None
```
I think it might be cleaner if this were an `else` block for the next `if` statement -- that way, this field only has a single value within this function, rather than having one set and then overwritten.

sahil839 replied:
Updated.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -519,16 +519,19 @@ paths:
 
-                                        Note: This event is only visible to admins.
+                                        **Changes**: Prior to Zulip 7.0 (feature level 160), this event was sent
+                                        to the target user only and only when delivery email of the user was changed.
```
"was sent only to organization administrators and the affected user, and this
event would only be triggered by changing the affected user's delivery email."

sahil839 replied:
This was actually sent to only affected user as per the [code here](https://github.com/zulip/zulip/blob/a6abf959bb68bcc6275ce94601a1d4ab21234319/zerver/actions/user_settings.py#L69) and the documentation was incorrect.

Also mentioned about `None` case.

timabbott commented on static/js/people.js:
```diff
@@ -1289,7 +1289,7 @@ function safe_lower(s) {
-    const email = settings_data.email_for_user_settings(person);
+    const email = person.delivery_email;
 
     return safe_lower(person.full_name).includes(value) || safe_lower(email).includes(value);
```
Hmm, this conditional seems like it won't enjoy `delivery_email` being `null`?

sahil839 replied:
This works because `safe_lower` takes care of this by using empty string for such case.

timabbott replied:
OK; I'm not sure I like that from a readability standpoint, but it's a preexisting quirk of this code path, and this PR is already complex enough.

timabbott commented on static/templates/settings/deactivated_users_admin.hbs:
```diff
@@ -13,7 +13,7 @@
             <thead class="table-sticky-headers">
                 <th class="active" data-sort="alphabetic" data-sort-prop="full_name">{{t "Name" }}</th>
-                <th {{#if can_sort_by_email}}data-sort="email"{{/if}}>{{t "Email" }}</th>
+                <th {{#if allow_sorting_deactivated_users_list_by_email}}data-sort="email"{{/if}}>{{t "Email" }}</th>
```
Why does this need a different structure than what we have in the normal users list?

sahil839 replied:
For normal users list, there will always be atleast one email (that of the user itself) which will be visible and thus we allow sorting.

timabbott replied:
OK -- I think what I'd expect is for the `data-sort` hover option to be available even with an empty table. But probably there's no strong reason to remove this logic if you've already got it "working".

timabbott commented on zerver/migrations/0425_migrate_to_user_level_email_address_visibility_setting.py:
```diff
@@ -0,0 +1,36 @@
+    for realm in realms:
+        UserProfile.objects.filter(realm=realm, is_bot=False).update(
+            email_address_visibility=realm.email_address_visibility
+        )
```
Maybe add a comment noting that database-leverl default value, for bots, of `EMAIL_ADDRESS_VISIBILITY_EVERYONE` is correct.

BTW, can you double-check that bots get `EMAIL_ADDRESS_VISIBILITY_EVERYONE` during account creation even if the realm's new user defaults setting is something different?

sahil839 replied:
> Maybe add a comment noting that database-leverl default value, for bots, of EMAIL_ADDRESS_VISIBILITY_EVERYONE is correct.

Aded comment.

Also, the `email_address_visibility` was not set to `EVERYONE` for bots, updated it now, so thanks for catching this.

timabbott replied:
Great, thanks!

gnprice reviewed (commented):
Thanks @sahil839! A few comments below. They're all on the documentation; if the code already behaves the way I hope, then they may all be purely docs issues.

There's also an open [chat thread](https://chat.zulip.org/#narrow/stream/378-api-design/topic/email.20address.20visibility/near/1498267) which corresponds to the above review thread https://github.com/zulip/zulip/pull/22994#discussion_r974173189 .

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -9095,6 +9150,31 @@ paths:
+            - 1 = everyone
+            - 2 = members only
+            - 3 = administrators only
+            - 4 = nobody (though note that administrators can change this setting).
```
No, missed when copying the description for realm setting. Fixed it now. Thanks.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -7189,11 +7240,15 @@ paths:
+
+                          **Changes**: Prior to Zulip 7.0 (feature level 160), this field was present only
+                          when `email_address_visibility` was restricted and you had access to real email address.
+                          Now, this field is present always with value being `None` if you do not have access to
```
Yes, the latter is correct. Noticed this while doing the changes Tim suggested above. Fixed this now.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -7189,11 +7240,15 @@ paths:
-                          limited and you are an administrator with access to real email addresses
-                          under the configured policy.
+                          The user's real email address. This field is present only if you have access
+                          to real email addresses under the [configured policy](/help/restrict-visibility-of-email-addresses).
```
This line is updated as you pointed out below.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -504,16 +519,19 @@ paths:
                                       description: |
-                                        When the delivery email of a user changes.
+                                        When the delivery email of a user changes or `email_address_visibility`
+                                        setting of a user changes.
```
Correct. Updated.

sahil839 commented:
I have updated the PR. Posted replies to the comments above. Have also updated the documentation as suggested above, but would be good to have a look at once again if it needs any change.

timabbott commented:
I'm waiting to attempt to review this further and try to integrate this on the API detail of how `client_gravatar` and `email_address_visibility` interact to be resolved, since I'm guessing I won't have material further comments once that's handled.

sahil839 commented on zerver/actions/user_settings.py:
```diff
@@ -488,19 +530,7 @@ def do_change_user_setting(
-                active_user_ids(user_profile.realm_id),
-            )
-        )
+        send_timezone_update_event_to_all_users(user_profile)
```
This leads to different events being sent to different emails based on email access. `email` field is already marked as `Depreceated` in the docs, so we can probably try to avoid this change. One workaround can be we do not update the user's `email` field in `apply_event` code using the email value passed in this event, since we already have different events to update email.

sahil839 commented:
Updated the PR as discussed on CZO. There are not much changes from previous state.
- Removed the `email_address_visibility` field from event objects and instead we do a database query in `apply_event` code to get the setting.
- The order of commits have changed and commit for `Always passing delivery_email in user objects` is now first commit. I changed the commit order while working to set `email` to real email for visibility setting other than `EVERYONE`, so have kept it as it is now.
- Documentation changes are same as were done for the previous round of feedback.

sahil839 commented:
Have not renamed `reset_emails_in_zulip_realm` as discussed before in this PR. I think would be better to do it separately, since this PR is already large enough.

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,18 @@ format used by the Zulip server that they are interacting with.
+  [`GET /users/me`](/api/get-own-user) and [`GET /events`](/api/get-events):
+  `delivery_email` field is passed in user objects always with value being `None`
+  if user does not have access to real email and also when `email_address_visibility`
+  is set to `EMAIL_ADDRESS_VISIBILITY_EVERYONE`.
```
Hmm, it seems like the API would be simpler if it was just identical to `email` if `EMAIL_ADDRESS_VISIBILITY_EVERYONE`? I think a correct client could "Display an email for a user if and only if `delivery_email` is not None) -- and never display the fake `email` field except as required for legacy API details like the search box that we're hoping to eliminate.

sahil839 replied:
Ohh, the way it is written here is confusing I guess. We set both `email` and `delivery_email` to real email for `EMAIL_ADDRESS_VISIBILITY_EVERYONE`. I wanted to write that previoulsy there was no `delivery_email` field in the object for `EMAIL_ADDRESS_VISIBILITY_EVERYONE`, but now there is. I agree that the way it is phrased makes it confusing.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -504,18 +504,30 @@ paths:
-                                        Note: This event is only visible to admins.
+                                        **Changes**: Prior to Zulip 7.0 (feature level 162), this event was sent
+                                        only to the affected user, and this event would only be triggered by
+                                        changing the affected user's delivery email.
```
I think this is inaccurate -- previously it would be sent both to the affected user and all organization admins, right?

sahil839 replied:
As per the code [here](https://github.com/zulip/zulip/blob/4dfe3d36e0c4373ea7f30daa894a4f195330e49c/zerver/actions/user_settings.py#L69), it was sent to affected user only. I guess the documentation here was incorrect, unless I am missing something.

timabbott replied:
Cool, didn't know that :)

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -7189,11 +7201,17 @@ paths:
+                          The user's real email address. This value will be `None` if you cannot
+                          access this user's real email address. For bots, this field is always
+                          set to the real email of the bot, since everyone can access the bot's
+                          email irrespective of `email_address_visibility` setting.
```
I think this would be better expressed as "because bot users always have EMAIL_ADDRESS_VISIBILITY_EVERYONE`? I checked and bots can't change `.email_address_visibility` via the API, so that is an invariant, and I think it's simper to explain it as such here, even if we short-circuit this in various code paths with an `is_bot` check.

timabbott commented on zerver/tests/test_events.py:
```diff
@@ -2633,8 +2633,11 @@ def do_set_realm_property_test(self, name: str) -> None:
+                # one delivery_email update event is sent for 9 users (all human
+                # users except the user themselves) + one email update event for
+                # each of the 10 human users in the realm.
+                num_events = 20
```
I guess this is problematically high (and thus a DoS risk), but that's OK because we're deleting this later in the PR; even the 1 per user was a problem. I found the existing issue for this problem and closed it: https://github.com/zulip/zulip/issues/14600.

timabbott commented on zerver/models.py:
```diff
@@ -1606,6 +1606,25 @@ class UserBaseSettings(models.Model):
 
+    # Who in the organization has access to users' actual email
+    # addresses.  Controls whether the UserProfile.email field is the
+    # same as UserProfile.delivery_email, or is instead garbage.
```
Let's replace "is instead garbage." with "is instead a fake generated value encoding the user ID and realm hostname."

Maybe also saying "the deprecated UserProfile.email field" would be good.

sahil839 replied:
Are we okay to mention "depreceated" just here in the comment and not in API docs.

timabbott commented on zerver/lib/bulk_create.py:
```diff
@@ -34,7 +34,12 @@ def bulk_create_users(
+        realm_user_default = RealmUserDefault.objects.get(realm=realm)
+        email_address_visibility = realm_user_default.email_address_visibility
+    else:
+        email_address_visibility = UserProfile.EMAIL_ADDRESS_VISIBILITY_EVERYONE
```
Can you add a comment "# There is no privacy motivation for limiting access to bot email addresses, so we hardcode them to EMAIL_ADDRESS_VISIBILITY_EVERYONE."

sahil839 replied:
Updated this,

timabbott commented on zerver/lib/create_user.py:
```diff
@@ -154,6 +157,14 @@ def create_user(
     enable_marketing_emails: Optional[bool] = None,
 ) -> UserProfile:
+    if bot_type is None:
+        realm_user_default = RealmUserDefault.objects.get(realm=realm)
```
Should we fetch this outside the `if` statement? I know it wastes a query when creating bots, but I don't like the control flow structure where we fetch it here conditionally and use it under the same condition below; just feels hard to analyze.

sahil839 replied:
Yeah, I did this for saving one query but I agree it is better to move the query before if statement. Updated it.

timabbott commented on zerver/models.py:
```diff
@@ -2062,9 +2062,7 @@ def allowed_bot_types(self) -> List[int]:
-        if self.realm.email_address_visibility == Realm.EMAIL_ADDRESS_VISIBILITY_EVERYONE:
-            return True
-        if self.is_bot:
+        if self.email_address_visibility == UserProfile.EMAIL_ADDRESS_VISIBILITY_EVERYONE:
```
Maybe worth a small comment noting that bots always have `EMAIL_ADDRESS_VISIBILITY_EVERYONE`?

sahil839 replied:
Added it.

timabbott commented on static/js/settings_config.ts:
```diff
@@ -123,11 +123,11 @@ export const email_address_visibility_values = {
+    },
     moderators: {
         code: 5,
         description: $t({defaultMessage: "Admins and moderators"}),
```
Would it be a good idea to align with the numbers for given groups used in `common_policy_values`? Or I guess `invite_to_realm_policy_values`, as the one with a `nobody` value? I think that to do that, we'd mostly just need to use the more standard values in `models.py` and map the IDs a bit differently in the migration file...

I guess the main question is whether that sort of alignment might make our life a bit easier when migrating to this being a "groups" based setting in the future, but maybe it doesn't matter and it's better to not mess with things... and I think the answer is no, in that we're just going to map a given value onto a given group using the `EMAIL_ADDRESS_VISIBILITY_EVERYONE` type names? But I'll leave it to you to decide -- we only have two settings today that don't share the meanings of which `code` goes with which group with the `common_policy_values` structure -- this one and `wildcard_mention_policy` -- and this is a basically free opportunity to make this one work the same as all the others.

As a side note, there's the question of whether we want to be migrating to a role dropdown rather than a "pick a group" setting here. I think the answer is probably we should merge this as written, because the conditionals of the form "Is this EMAIL_ADDRESS_VISIBILITY_EVERYONE" will be hard to port without doing a database query, and that seems like enough of a rabbit hole that we're probably better off having that be an independent migration in the future. But wanted to raise this in case you disagree, since you're probably going to be the one doing such a migration down the line.

sahil839 replied:
Regarding the option values, I think we would just require to map the values in migration file and I don't think it would matter much there for what values we have for this and other settings.

And I think it's better to do this as role thing and not groups. We still have some things to figure out for groups based settings regarding code structure and this PR is already big enough.

timabbott replied:
Yeah, I agree on both points, resolving.

timabbott commented on static/styles/settings.css:
```diff
@@ -1476,7 +1476,9 @@ $option_title_width: 180px;
+#user_email_address_visibility,
+#realm_email_address_visibility {
     width: 325px;
 }
```
Just a side comment -- we almost certainly want to look at replacing this rule with a `.settings-select` class that we apply to all of our settings select elements, so we stop having to fiddle with this CSS file whenever we add a new dropdown. Definitely a thing for a separate PR, though.

sahil839 replied:
We discussed this before in some PR a while ago. I intend to start a discussion in #design.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -3770,6 +3770,12 @@ paths:
+                                It was replaced by a [user setting](/api/update-settings) with a
+                                [realm user default](/api/update-realm-user-settings-defaults), with the encoding of different
+                                values preserved. Clients can support all versions by supporting the current API and treating
+                                every user as having the realm's email address visibility value.
```
I guess the fact that the encoding is the same as it was before makes this a lot easier to write -- that perhaps argues for not standardizing to the `common_policy_types` encoding with this migration.

timabbott commented:
Cool, this is great work. I posted a bunch of tactical comments mostly on API documentation, plus https://github.com/zulip/zulip/pull/22994#discussion_r1099519185 which revisits a few details of the database model worth at least making an explicit decision on. My current opinion is that probably all things considered, it's better to integrate this as written on both points, but wanted to make sure you're thinking about it too.

sahil839 commented on static/js/settings_config.ts:
```diff
@@ -123,11 +123,11 @@ export const email_address_visibility_values = {
-    //     code: 2,
-    //     description: $t({defaultMessage: "Members and admins"}),
-    // },
+    // Backend support for this configuration is not available yet.
```
No. Removed it.

sahil839 commented:
Updated the PR. Made the documentation changes and changes related to comments. Also replied to the other API discussion in the comments above.

One more change that I have done is that removed `target_is_bot` parameter from `can_access_delivery_email`, since we set the visibility to `EVERYONE` for bots.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -9105,6 +9117,31 @@ paths:
+            The [policy][permission-level] this user has selected for which other
+            users in this organization can see the real email address of this user.
+
+            - 1 = everyone
+            - 2 = members only
+            - 3 = administrators only
+            - 4 = nobody
+            - 5 = moderators only
```
nit: I think title-case (like we have in the doc for the old realm setting) is a bit more readable in this sort of table:

```suggestion
            - 1 = Everyone
            - 2 = Members only
            - 3 = Administrators only
            - 4 = Nobody
            - 5 = Moderators only
```

(Similarly elsewhere.)

sahil839 replied:
Updated for this. We can make this consistent at other places too, we still use lower cases in some places.

timabbott replied:
I think it's not a big priority to fix that, in that we intended to replace almost all of these settings with groups, but maybe it's worth a sweep since it's not hard and we will still mention there in **changes** API documentation about previous states.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -7189,11 +7192,20 @@ paths:
-                          under the configured policy.
+                          The user's real email address. This value will be `None` if you cannot
+                          access this user's real email address. For bot users, this field is
+                          always set to the real email of the bot, because bot users always have
+                          `email_address_visibility` as `EMAIL_ADDRESS_VISIBILITY_EVERYONE`.
```
commit-structure nit: there's a few changes in this commit:

588c45f3b settings: Add backend code for using user email_address_visibility setting.

that look like they were meant to be squashed into the previous one. (They're wording changes to the API docs, but this commit doesn't otherwise touch the API.)

sahil839 replied:
Yeah, so the change to always show emails for bots is done in the previous commit `users: Always pass "delivery_email" field in user objects` but as Tim suggested [here](https://github.com/zulip/zulip/pull/22994#discussion_r1099493434) to write it in different way and we actually use the user-level setting in commit ` settings: Add backend code for using user email_address_visibility setting.`  I am not sure whether we are fine with such type of changes but thought it would be misleading to mention about the user-level setting when we are actually not using it.

gnprice reviewed (commented):
Thanks @sahil839! One substantive comment below, and several small bits of wording in the docs.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -504,18 +504,30 @@ paths:
                                     - type: object
                                       additionalProperties: false
                                       description: |
-                                        When the delivery email of a user changes.
+                                        When the delivery email of a user changes or access to the delivery email
+                                        changes for you due to change in `email_address_visibility` setting.
```
The event is only sent when user can see the email. I am not sure how to explain that clearly here. Can we have something like - `When the delivery email of a user, whose email you can access, changes or access ......` ?

timabbott replied:
I think the right way to explain it is "Event sent when the value of the user's `delivery_email` as visible to you changes, either due to the email address changing or your access to the user's email via their `email_address_visibility` policy changing."

As a side note, I bet we have a bug where if your role changes, your don't get `email_address_visibility` events for users for whom your access to their email address changes. I think we should leave addressing that as a follow-up; we already have various other things that don't quite live-update perfectly after role changes. Opened https://github.com/zulip/zulip/issues/24350 to track this.

timabbott replied:
I see this change was applied.

sahil839 commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,26 @@ format used by the Zulip server that they are interacting with.
+  `delivery_email` field is passed in user objects always including the case
+  when `email_address_visibility` is set to `EMAIL_ADDRESS_VISIBILITY_EVERYONE`
+  with value being `None` if user does not have access to real email. For bot
+  users, `delivery_email` field is always set to real email.
```
Thanks, updated this. Will be careful next time.

sahil839 commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,26 @@ format used by the Zulip server that they are interacting with.
+  /settings`](/api/update-settings), [`PATCH
+  /realm/user_settings_defaults`](/api/update-realm-user-settings-defaults): Added
+  user `email_address_visibility` setting which will replace the existing realm
+  `email_address_visibility` setting.
```
Updated.

sahil839 commented:
@gnprice updated the docs and replied to the two comments above.

timabbott commented:
Cool I posted a small comment, but my plan is to try to integrate this to `main` tomorrow.

sahil839 commented:
Done the documentation change as suggested above.

timabbott commented:
Merged, after amending the last commit to close two of the issues related to this feature. Huge thanks for all the hard work on this @sahil839!

#16920 is still open.
